### DISCUSSION TITLE
JasperReport e Angular

### DISCUSSION THREAD

**Discussion by guilhermecostalopes** [frontendbr/forum#1278] [Perguntas e Respostas]
Tenho um **jasper**. chamado e criado pelo método abaixo. Não dá erro no JAVA. Entendo que está ok.

Como fazer para o usuário fazer o download após ser gerado ou como abrir em um popup o relatório gerado ?

    package br.eti.netsoft.authSeguranca.relatorio;

    import java.io.Serializable;
    import java.util.HashMap;
    import java.util.Map;

    import javax.servlet.http.HttpServletRequest;

    import net.sf.jasperreports.engine.JRException;
    import net.sf.jasperreports.engine.JasperRunManager;

    import org.springframework.http.MediaType;
    import org.springframework.web.bind.annotation.CrossOrigin;
    import org.springframework.web.bind.annotation.PostMapping;
    import org.springframework.web.bind.annotation.RequestMapping;
    import org.springframework.web.bind.annotation.RestController;

    import br.eti.netsoft.authSeguranca.download.DownloadCommand;
    import br.eti.netsoft.authSeguranca.download.ResponseData;
    import br.eti.netsoft.exception.BusinessException;

    @RestController
    @RequestMapping("/admin/seg/relatorio")
    @CrossOrigin(origins = "*")
    public class RelatorioJasper implements DownloadCommand, Serializable {

    	private static final long serialVersionUID = -674056638665799813L;

    	private static final String RELATORIOS = "/relatorios/";
    	private static final String TESTE = "grupo";
    	private static final String JASPER = ".jasper";

    	public static final String NOME_ARQUIVO_RECEBIMENTO_PROVISORIO = "Teste.pdf";
    	public static final String APPLICATION_PDF = "application/pdf";

    	@PostMapping(value = "/grupo", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
    	@Override
    	public ResponseData execute(HttpServletRequest req)
    			throws BusinessException {
    		Map<String, Object> parameters = new HashMap<>();
    		byte[] pdf = null;
    		try {
    			pdf = JasperRunManager.runReportToPdf(
    					this.getClass().getClassLoader()
    							.getResourceAsStream(RELATORIOS + TESTE + JASPER),
    					parameters);
    		} catch (JRException e) {
    			e.printStackTrace();
    		}
    		ResponseData rd = new ResponseData();
    		rd.setData(pdf);
    		rd.setFileName(NOME_ARQUIVO_RECEBIMENTO_PROVISORIO);
    		rd.setContentDisposition("inline");
    		rd.setContentType(APPLICATION_PDF);
    		return rd;
    	}
    }

no console do browser vejo este retorno:
    contentDisposition: "inline"
    contentType: "application/pdf"
    data: "JVBERi0xLjUKJeLjz9MKMSAwIG9iago8PC9GaWx0ZXIvRmxhdGVEZWNvZGUvTGVuZ3RoIDI5Pj5zdHJlYW0KeJwr5HIK4TI2U7AwMFMISeFyDeEK5OICADEzBC0KZW5kc3RyZWFtCmVuZG9iagozIDAgb2JqCjw8L1RhYnMvUy9Hcm91cDw8L1MvVHJhbnNwYXJlbmN5L1R5cGUvR3JvdXAvQ1MvRGV2aWNlUkdCPj4vQ29udGVudHMgMSAwIFIvVHlwZS9QYWdlL1Jlc291cmNlczw8L0NvbG9yU3BhY2U8PC9DUy9EZXZpY2VSR0I+Pi9Qcm9jU2V0IFsvUERGIC9UZXh0IC9JbWFnZUIgL0ltYWdlQyAvSW1hZ2VJXT4+L1BhcmVudCAyIDAgUi9NZWRpYUJveFswIDAgNTk1IDg0Ml0+PgplbmRvYmoKMiAwIG9iago8PC9LaWRzWzMgMCBSXS9UeXBlL1BhZ2VzL0NvdW50IDEvSVRYVCgyLjEuNyk+PgplbmRvYmoKNCAwIG9iago8PC9UeXBlL0NhdGFsb2cvUGFnZXMgMiAwIFIvVmlld2VyUHJlZmVyZW5jZXM8PC9QcmludFNjYWxpbmcvQXBwRGVmYXVsdD4+Pj4KZW5kb2JqCjUgMCBvYmoKPDwvTW9kRGF0ZShEOjIwMTgxMTE0MTkwMTEyLTAyJzAwJykvQ3JlYXRvcihKYXNwZXJSZXBvcnRzIExpYnJhcnkgdmVyc2lvbiA2LjYuMCkvQ3JlYXRpb25EYXRlKEQ6MjAxODExMTQxOTAxMTItMDInMDAnKS9Qcm9kdWNlcihpVGV4dCAyLjEuNyBieSAxVDNYVCk+PgplbmRvYmoKeHJlZgowIDYKMDAwMDAwMDAwMCA2NTUzNSBmIAowMDAwMDAwMDE1IDAwMDAwIG4gCjAwMDAwMDAzMzMgMDAwMDAgbiAKMDAwMDAwMDExMCAwMDAwMCBuIAowMDAwMDAwMzk2IDAwMDAwIG4gCjAwMDAwMDA0ODcgMDAwMDAgbiAKdHJhaWxlcgo8PC9JbmZvIDUgMCBSL0lEIFs8ODUzMzMxY2Q1NWNkOWExYWQzNDIzNDQ1OWM1NTZkMzk+PDFkMGY3ZGU3ZjgzMmVjNjk1OThjN2I0OTc2OTYxYTZiPl0vUm9vdCA0IDAgUi9TaXplIDY+PgpzdGFydHhyZWYKNjU0CiUlRU9GCg=="
    fileName: "Teste.pdf"

![console](https://user-images.githubusercontent.com/24395029/48512498-c1b1b500-e840-11e8-80a4-5c9161e77ecd.png)

**Comment by iamfelipemattos** [comment]
@guilhermecostalopes voce esta usando qual versao do angular?

sobre como fazer o download é simples. seguinte, voce vai precisar pegar a response do seu endpoint que retorna informações do seu arquivo, na qual voce mostrou no print. até ai tudo bem?
Após isso voce voce vai precisar fazer o download do arquivo, pra isso, eu sempre utilizei o [file saver](https://www.npmjs.com/package/file-saver), bem simples de implementar, dai é só implementar um metodo e chama-lo no seu botao de salvar.

A questao de abrir em um popup o relatório gerado, voce precisara criar uma modal ou utilizar alguma lib que a disponibiliza e entao chamar o endpoint que traz o relatorio gerado, no retorno da response voce chama seu metodo que abre a modal e passa pra ela os dados da response que foi retornada e exibe no html.

qualquer coisa pinga nois ai. tmj
